packages/yew/src/dom_bundle/btag/listeners.rs

```diff
@@ -141,24 +142,31 @@ struct GlobalHandlers {
     registered: Vec<(ListenerKind, EventListener)>,
 }
 
-impl GlobalHandlers {
+impl HostHandlers {
+    fn new(host: HtmlEventTarget) -> Self {
+        Self {
+            host,
+            handling: HashSet::default(),
+            #[cfg(test)]
+            registered: Vec::default(),
+        }
+    }
+
     /// Ensure a descriptor has a global event handler assigned
-    fn ensure_handled(&mut self, desc: EventDescriptor) {
+    fn ensure_handled(&mut self, root: &BSubtree, desc: EventDescriptor) {
         if !self.handling.contains(&desc) {
             let cl = {
                 let desc = desc.clone();
-                BODY.with(move |body| {
-                    let options = EventListenerOptions {
-                        phase: EventListenerPhase::Capture,
-                        passive: desc.passive,
-                    };
-                    EventListener::new_with_options(
-                        body,
-                        desc.kind.type_name(),
-                        options,
-                        move |e: &Event| Registry::handle(desc.clone(), e.clone()),
-                    )
-                })
+                let options = EventListenerOptions {
+                    phase: EventListenerPhase::Capture,
+                    passive: desc.passive,
+                };
+                EventListener::new_with_options(
+                    &self.host,
```

Comment:
Will events be dispatched multiple times if the host of a portal is also the child of the root of the actual application?

We have a different recommendation than React in terms of where the application should be mounted.
React recommends mounting the application into a `#root` container which means that any portal hosts should not be a child of the root application.

However, Yew recommends to mount the application to `document.body` which means that the host element of any event dispatchable portals will be a child element of the root.

Reply:
Events should be dispatched only once, refer to [the method for figuring out on which subtree to do this](https://github.com/yewstack/yew/pull/2510/files#diff-2734bc9f7218f3cfe78a6b6fe7236f883b0be2df1434eb64201f924c1eb0b50aR289-R351), all other subtrees ignore the event. I'm almost certain that even react has the same issue to solve, since you can portal into other parts of the (controlled) dom tree with refs.

Secondly, elements in a shadow-dom: I'd have to investigate a bit more. I agree that only one event should be handled, but I remember native event propagation to be a bit hairy, i.e. I think it changes the target when bubbling across the shadow-dom host, but I might be misremembering things.

Reply:
Well, I think event handling through ShadowDOM is broken.

![broken](https://user-images.githubusercontent.com/11693215/159127621-ab0307bb-c2da-4048-a486-bb92ba56f0e8.gif)

Reply:
The example in the PR works fine for me. In any case, closed shadow roots definitely need more work and docs. Open shadow roots should work fine. The reason being the retargeting: we'd need an event handler inside the shadow root to catch the "actual" event with the specific targeted element inside. Any outer handler won't see internal details. Which means, apparently, to either install handlers for *all* events pre-emptively or do some costly registration on all subtrees :/ I think that's what react does, anyway.

The other option is to find a design that embraces closed shadow roots as an abstraction boundary, i.e. only support open shadow roots for now.

Reply:
I didn't mention in the previous comment, I added an additional button that also registers an onclick handler outside of the shadowroot which is why it's broken.

I have also tested the the open mode, which does not seem to work either.

Reply:
Okay, I'm back from experimenting. Even in react, if using a closed shadow dom, click events that happen inside, but have a handler installed outside, will *fire multiple times* (once for each retargeting, with all different targets). Which makes me feel more comfortable with fixing the problem of event retargeting in future PRs and just documenting/using *open* shadow dom for now.

There's the non-standard [`explicitOriginalTarget`](https://developer.mozilla.org/en-US/docs/Web/API/Event/explicitOriginalTarget), but I'd stick to the first entry of [`composedPath`](https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath) for dispatching the event on, for now. The whole thing is a bit messy.

Would it make sense to introduce an element such as `<EventBoundary>` that stops events from bubbling through them? I think that could be useful to solve the issues with multiple events from closed shadow roots (once the current impl is fixed).

Reply:
We are talking about 2 different issues.
![broken2](https://user-images.githubusercontent.com/11693215/159255865-9aa00630-8acd-4cbb-8fd9-29da7a17915b.gif)

The issue I mentioned does not work in open mode of shadow dom either.
And no event listeners are notified for an onclick event (not notified for multiple times for the same event).

Reply:
> Would it make sense to introduce an element, maybe called <EventBoundary>, that stops events from bubbling through it?

After some further reading, I think there is a way to detect whether a portal is being rendered into a ShadowDom and the parent element of the Portal (in terms of the `parent_element` parameter passed at the time of attaching.) is a parent element of the portal host after passing through shadow host.

(You can search through `portal_host.parent_node` to see if a shadow root presents and then in turn searches if the host of the shadow root is (or a child node of) the `parent_element`. We only need to do this once at the time the Portal is created as this condition can never change after creation even if the portal is shifted afterwards.)

If we determined that the following are true:
- the Portal host is rendered into a shadow dom.
- the host of the shadow dom (or after many layers of shadow dom) is contained by the parent element of the portal.
- the `Event`'s `composed` property is `true`.

then it means that another event will be dispatched eventually so we can discard the event at the Portal.

This means that we also follow the native event retargeting behaviour correctly.

Reply:
> We are talking about 2 different issues.
> 
> The issue I mentioned does not work in open mode of shadow dom either. And no event listeners are notified for an onclick event (not notified for multiple times for the same event).

Thee current impl is bugged either way, hence why I converted the PR back to draft mode. All of this can hopefully be converted to test cases ;) Stopping/retargeting event bubbling at shadow-dom hosts sounds like a decent idea, until the following test case:

```
<div>
  <ShadowDiv>
    <VNode::VRef(<div id="trgt" />)/>
  </ShadowDiv>
  <div {onclick}>
    {create_portal(
      <button>{"Click me"}</button>,
     #trgt,
    )}
  </div>
</div>
```

You see a portal into a shadow dom, that is not a physical ancestor of the shadow root. Should the `onclick` handler receive the original,inner or a retargeted,outer event? If the former, it sees a different event than a handler installed on the outer-most `<div>`. If the latter, who would be responsible for triggering the handler? Bubbling the retargeted event (from `<ShadowDiv>`) never reaches the listening div.

Is that example contrived? Yes. But I also don't want to commit to wrong assumptions about the component hierarchy.

Reply:
The resolution I've gone for is to not respect event retargeting, for now. Open shadow doms are handled correctly now, closed shadows lead to two events, one bubbling from inside and one bubbling from the host element. These may or may not eventually lead to the same virtual element. In any case, it's also a behaviour other frameworks get "wrong" or buggy, for example facebook/react#24136 (in this framework, the event is even duplicated if the shadow root is open).

If anybody comes up with an actual use case to follow; I think a useful think to do would be to have an option to stop event bubbling at portals.

Reply:
This has been fixed after latest commit.